Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+    }
+
+    /// Yields a sequence of new elements to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` is writable the sequence will get forwarded to the ``NIOAsyncWriterDelegate`` immediately.
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter contentsOf: The sequence to yield.
+    @inlinable
+    public func yield<S: Sequence>(contentsOf sequence: S) async throws where S.Element == Element {
+        try await self._storage.yield(contentsOf: sequence)
+    }
+
+    /// Yields an element to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` is writable the element will get forwarded to the ``NIOAsyncWriterDelegate`` immediately.
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter element: The element to yield.
+    @inlinable
+    public func yield(_ element: Element) async throws {
+        try await self.yield(contentsOf: CollectionOfOne(element))
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    @inlinable
+    public func finish() {
+        self._storage.finish(with: nil)
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter failure: The failure indicating why the writer finished.
+    @inlinable
+    public func finish(with failure: Failure) {
+        self._storage.finish(with: failure)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to toggle the writability of the ``NIOAsyncWriter``.
+    public struct Sink {
```

Comment:
What's the benefit of this type? Ensure that a consumer of the `NIOAsyncWriter` can only call `setWritability` on it? In my mind the `OutboundChannelHandler` implements the Delegate protocol and holds a reference to the writer. `OutboundChannelHandler` can then easily toggle writability on the `NIOAsyncWriter`.

Reply:
The benefit of this type is that the producer and the delegate are holding two separate types. This allows us to make assumption of when the `NIOAsyncWriter` actually gets deallocated since this means we can tell the delegate that we terminated. I stated in the doc comments of `makeWriter` that you should hold the `Sink` in the place where you consume the writes and the `AsyncWriter` in the async world.